### ISSUE TITLE
Generics for messages are part of structure

### ISSUE THREAD

**Issue by insanitybit** [insanitybit/derive_aktor#12]
Unfortunately, because the message queues must exist in the struct of the actor, all generics for all methods are exposed at that level. This sucks, since if my Actor struct A has to hold a cache actor B, All generics for B are lifted to A. This leads to a huge explosion in type parameters, and they have to be written out.

On top of this, given that return values will use closures, actor messages will always have multiple generic types for return closures. For every method that returns it's +1 generic type.

The solution may be to use unsafe. If we validate the types are Send + 'static, we can safely transmute them to raw bytes or something like that, send across the queue, and then transmute them to their proper type. Because the types are known ahead of time this should be safe - we'll still have enum tags telling us which method is being routed to. This should also have 0 cost.

Motivating example:

```rust
struct CacheActor<FGet, FPut> {
    message_box: Queue<CacheMessage<FGet, FPut>>
}

struct DBActor<DBGet, DBPut, FGet, FPut> {
    message_box: Queue<DBMessage<DBGet, DBPut, GFet, FPut>>
}

```

The Get/Put corresponds to the closure for a 'get' or 'put' operation, for a fake database/ cache actor. Because the DBActor holds onto the CacheActor, it must expose its types, as far as I can tell.

There may be another way around this that I'm not seeing.

**Comment by insanitybit** [comment]
I believe the way around this is with some unsafe and transmute.

Messages will now look more like this:
```rust
FooMessage {
    Bar([u8; 24]),
    Baz([u8; 24])
}
```

for the code

```rust
impl Foo {
    fn bar<F: FnOnce()>(u32, Box<F>) {}
    fn baz<F: FnOnce()>(u32, Box<F>) {}
}
```

Notice, no generics.  To call 'bar' or 'baz' you must satisfy the type constraints of those methods. There's no need for our Queue to be aware of the types as well.

We can transmute the arguments into a tuple and store it in the enum. This way our struct doesn't require generic parameters just to talk about types that already exist.

**Comment by insanitybit** [comment]
So because there is no way to, at compile time, determine the size of a type, we have to just throw some memory away and perform some copies. This is just how it'll be for now.

So every variant will be more like:

FooMessage {
    Bar([u8; 1024]),
    Baz([u8; 1024])
}

And then the transmute will be a transmute_copy, and will require an assert on the size of args before copying.

This means you're limited to 1024 bytes (or some other number, maybe 512) and all extra space is just wasted.

**Comment by insanitybit** [comment]
I tried this out. There are some problems.

Here is the output of the code I tried out after macro expansion:

https://play.integer32.com/?gist=b16e4f442a264363d85fcd77062cb427&version=undefined
https://gist.github.com/anonymous/b16e4f442a264363d85fcd77062cb427

The compiler complains that the route_msg method can't infer the generic type. This is unsurprising, really. It has no idea what the concrete type is and we have nothing telling it.

We need a way to tie these generic types together without exposing these generics to the top level actor.

The other issue is when transmuting 'drop' items:
https://www.reddit.com/r/rust/comments/6gftll/issue_with_type_inference_looking_for_validation/diq83hz/

By performing the transmute we copy a pointer to a buffer, without tracking it/ changing ownership. The old pointer goes out of scope, freeing the data, and accessing the new pointer is now UB. This one's easy to work around, though it seems to me like the issue is going to need to be solved differently as a whole.
